drivers/soft_uart/soft_uart.c

```diff
@@ -0,0 +1,323 @@
+
+    if (ctx->state_rx == STATE_RX_HIGH) {
+        ctx->byte_rx &= ~mask;
+        ctx->state_rx = STATE_RX_LOW;
+    } else {
+        ctx->byte_rx |= mask;
+        ctx->state_rx = STATE_RX_HIGH;
+    }
+}
+
+int soft_uart_init(soft_uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
+{
+    if (uart >= SOFT_UART_NUMOF) {
+        return UART_NODEV;
+    }
+
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    mutex_init(&ctx->lock);
+    mutex_init(&ctx->sync);
+    mutex_lock(&ctx->sync);
+
+    ctx->baud = cfg->timer_freq / baudrate;
+
+    if (cfg->timer_freq / ctx->baud - baudrate > 500) {
+        return UART_NOBAUD;
+    }
+
+    if (cfg->rx_pin == GPIO_UNDEF) {
+        rx_cb = NULL;
+    }
+
+#ifdef MODULE_SOFT_UART_MODECFG
+    ctx->data_bits = 8;
+    ctx->stop_bits = 1;
+    ctx->parity    = PARITY_NONE;
+#endif
+
+    ctx->rx_cb = rx_cb;
+    ctx->rx_cb_arg = arg;
+
+    ctx->state_rx = STATE_RX_IDLE;
+
+    if (cfg->tx_pin != GPIO_UNDEF) {
+        timer_init(cfg->tx_timer, cfg->timer_freq, _tx_timer_cb, (void*) uart);
+        gpio_write(cfg->tx_pin, !(cfg->flags & SOFT_UART_FLAG_INVERT_TX));
+        gpio_init(cfg->tx_pin, GPIO_OUT);
+    }
+
+    if (rx_cb) {
+        timer_init(cfg->rx_timer, cfg->timer_freq, _rx_timer_cb, (void*) uart);
+        timer_stop(cfg->rx_timer);
+        /* timer should fire at the end of the byte */
+        timer_set_periodic(cfg->rx_timer, 0, ctx->baud * (BITS_DATA(ctx) + BITS_PARITY(ctx) + 1),
+                           TIM_FLAG_RESET_ON_MATCH | TIM_FLAG_RESET_ON_SET);
+        gpio_init_int(cfg->rx_pin, GPIO_IN, GPIO_BOTH, _rx_gpio_cb, (void*) uart);
+    }
+
+    return 0;
+}
+
+#ifdef MODULE_SOFT_UART_MODECFG
+int soft_uart_mode(soft_uart_t uart, uart_data_bits_t data_bits, uart_parity_t parity,
+                   uart_stop_bits_t stop_bits)
+{
+    if (uart >= SOFT_UART_NUMOF) {
+        return UART_NODEV;
+    }
+
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    switch (data_bits) {
+    case UART_DATA_BITS_5:
+        ctx->data_bits = 5;
+        break;
+    case UART_DATA_BITS_6:
+        ctx->data_bits = 6;
+        break;
+    case UART_DATA_BITS_7:
+        ctx->data_bits = 7;
+        break;
+    case UART_DATA_BITS_8:
+        ctx->data_bits = 8;
+        break;
+    }
+
+    switch (parity) {
+    case UART_PARITY_NONE:
+        ctx->parity = PARITY_NONE;
+        break;
+    case UART_PARITY_EVEN:
+        ctx->parity = PARITY_EVEN;
+        break;
+    case UART_PARITY_ODD:
+        ctx->parity = PARITY_ODD;
+        break;
+    case UART_PARITY_MARK:
+        ctx->parity = PARITY_MARK;
+        break;
+    case UART_PARITY_SPACE:
+        ctx->parity = PARITY_SPACE;
+        break;
+    }
+
+    switch (stop_bits) {
+    case UART_STOP_BITS_1:
+        ctx->stop_bits = 1;
+        break;
+    case UART_STOP_BITS_2:
+        ctx->stop_bits = 2;
+        break;
+    }
```

Comment:
```suggestion
    ctx->stop_bits = 1;
    if (stop_bits == UART_STOP_BITS_2) {
        ctx->stop_bits = 2;
    }
```

This would also make sure that `ctx->stop_bits` is always initialized. (In C, an `enum` is just an integer of unspecified width and signess that is big enough to represent all defined constants. We can do something like `(enum weekday)42` and get a weekday that is not in the range [0;6].)

Reply:
I'll just add a `default` case so it doesn't look out of place with the other `switch` statements.